### DISCUSSION TITLE
Set MQTT connection loss retry time?

### DISCUSSION THREAD

**Discussion by Hacklzwerg** [arendst/Tasmota#14632] [Q&A] [answered]
Hello.
as new member in the community allow me to say: Tasmota is one of the best projects i've ever seen..

I have a humble question:
Is it possible to change the incrementing of the MQTT connection loss retry counter by 10 seconds (i.e. first attempt in 10 seconds, second attempt in 20 seconds, third in 30 seconds, etc.).
and set it so, that only a specified time (e.g. 10 seconds) elapses with each attempt and then the connection attempt is resumed?
(i.e. 1st attempt after 10 seconds, 2nd attempt after another 10 seconds, third again after 10 seconds etc etc...) 
Using Tasmota 10.1.0

Greets out of Lower Bavaria
Stefan

**Comment by barbudor** [comment]
Greetings from France

You can't change the Tasmota behavior to increment the timeout at each retry. This is a common practice when  a connection timeouts.
However you can change the base timeout with MqttRetry command
But nevertheless, without changing Tasmota code, on each failing retry the timeout will increase until it reaches 120 seconds.

**Comment by Hacklzwerg** [comment]
bonjour la france!

Well, thanks for your really fast answer.
I know that it is not possible to change this behavior at the moment..
and i know that maximum period is 120 secs..   
_Two minutes were nothing happens and the "Smart Home" isn't smart...._ 
But even whilst this is common practice, it may not apply or be satisfying to everyone (like me and some of my discerning clients).

Maybe it would be good to be able to set this behavior via "SetOption".

for example:
SetOptionXXX    Set MQTT retry period after Broker connection loss
                         0 = normal behavior 10s, 20s, 30s (default)
                         5..120 = seconds to next connection attempt (e.g. 15s, 15s, 15s...)

**Reply by Hacklzwerg** [reply]
moved to ideas   Stefan

**Comment by Jason2866** [comment]
Change this behaviour is doctoring on symptoms. Cure the reason for. Make your wifi and mqtt broker solid.

**Reply by Hacklzwerg** [reply]
With all due respect:
What's the point of this answer?
How do you want to know if my broker / WiFi is reliable or not? 
Could there be other reasons for making this suggestion?

**Reply by Jason2866** [reply]
If you have rc-2 messages you have a issue with wifi or mqtt broker.

**Reply by joba-1** [reply]
I think the point is wifi is an inherently unreliable medium and devices should be able to deal with it gracefully. Making wifi more reliable makes errors less likely, but can never avoid them completely.

**Reply by Hacklzwerg** [reply]
Well, there is a trivial reason for the rc-2 message...
The broker is intentionally switched off for a longer period of time, but the Tasmota devices remain connected to the powerline.
If the broker then is started up again, it can take up to two minutes for the devices to reconnect.
But it is desired that when the broker is online, the Tasmotas should also work again - and not only when two minutes have passed ...

**Reply by Jason2866** [reply]
The reason for enhancing the time is not to "hammer" the network/broker. When it is a planned switch of, take the 2 minutes extra needed in account.

**Reply by joba-1** [reply]
ok, so reconnect works but isnt fast enough for your usecase. Then my point is irrelevant here. Maybe you should describe better what your usecase to convince why your usecase needs this? That has gotten a lot of new features into tasmota already.

**Reply by barbudor** [reply]
@Hacklzwerg I'm not sure to see why these 2 minutes are critical for you.
Use where I see that would be a complete power cut and restore where the Tasmotas restart in 15 seconds while the server takes a few minutes to restart.
In that case, what is the criticity of the Tasmotas reconnecting to the server in less than 2 minutes after it is back on line ?
Everytime I restart my server because of a linux update, the MQTT broker is unavailable for 3 or 4 minutes. So what ?

**Reply by Hacklzwerg** [reply]
lol   :-)
Well... I personally don't mind either...
But have you ever tried to convince a very opinionated woman that things are just so how they are ???
The "queen of the palace" just wants that when the broker has booted up, she can switch lights and change the light colors straight to her heart's content. A very, very demanding custumer.
btw Broker is a Raspberry Pi4, doesn't take long to boot

**Reply by Jason2866** [reply]
Having demamding customer, needs special attention. If i where you, RasPi would be connected to a UPS and never be switched of when customer is at home 😁

**Comment by joba-1** [comment, marked as answer]
is there a possibility in current code to lure tasmota into an immediate mqtt retry? E.g. by changing the retry time via http request from pi to tasmota immediately after mqtt broker is online. That would be ideal, I guess

**Reply by Hacklzwerg** [reply]
oh... a very good approach.
I could have figured that out myself...
I'll try that right now..... 😁

**Reply by Hacklzwerg** [reply]
so have tried your suggestion.
Doesn't work. Seems the internal counter isn't reset with cmnd MqttRetry 10...
see timesstamps

```
14:10:29.526 MQT: stat/tasmota_xxxxxx/POWER = {"POWER":"OFF"}
14:11:25.483 MQT: Verbindungsversuch...
14:11:25.495 MQT: Verbindung fehlgeschlagen aufgrund von 192.168.XXX.XXX:XXXX, rc -2. Wiederversuch in 10 s
14:11:36.479 MQT: Verbindungsversuch...
14:11:36.486 MQT: Verbindung fehlgeschlagen aufgrund von 192.168.XXX.XXX:XXXX, rc -2. Wiederversuch in 20 s
14:11:57.491 MQT: Verbindungsversuch...
14:11:57.503 MQT: Verbindung fehlgeschlagen aufgrund von 192.168.XXX.XXX:XXXX, rc -2. Wiederversuch in 30 s
14:12:28.496 MQT: Verbindungsversuch...
14:12:28.503 MQT: Verbindung fehlgeschlagen aufgrund von 192.168.XXX.XXX:XXXX, rc -2. Wiederversuch in 40 s
14:13:09.476 MQT: Verbindungsversuch...
14:13:09.541 MQT: Verbindung fehlgeschlagen aufgrund von 192.168.XXX.XXX:XXXX, rc -2. Wiederversuch in 50 s
-->14:13:27.049 RSL: MQTTRETRY = {"MqttRetry":10}
14:13:32.403 RSL: STATE = {"Time":"2022-01-29T14:13:32","Uptime":"1T03:15:21","UptimeSec":98121,"Heap":27,.........}}
-->14:14:00.481 MQT: Verbindungsversuch...
14:14:00.497 MQT: verbunden
14:14:00.501 MQT: tele/tasmota_xxxxxx/LWT = Online (beibehalten)
14:14:00.536 MQT: stat/tasmota_xxxxxx/POWER = {"POWER":"OFF"}

**Reply by barbudor** [reply]
restart tasmota, it only takes 15 seconds

**Reply by Hacklzwerg** [reply]
Success!!

both approaches work 
@Joba-1 
your suggestion works if MqttRetry is set to another value than the default (10)
i set it to 11, so internal counter seems to be reset..
only have to set MqttRetry to default value 10 again, so that next attempt with 11 will work again...

@barbudor
your suggestion also works flawlessly

Thank you all for your help, I think my problem is solved here!!!! 

```
14:56:31.234 MQT: Verbindung fehlgeschlagen aufgrund von 192.168.XXX.XXX:XXXX, rc -2. Wiederversuch in 40 s
14:57:12.206 MQT: Verbindungsversuch...
14:57:12.213 MQT: Verbindung fehlgeschlagen aufgrund von 192.168.XXX.XXX:XXXX, rc -2. Wiederversuch in 50 s
---> 14:57:30.952 RSL: MQTTRETRY = {"MqttRetry":11}
14:57:42.206 MQT: Verbindungsversuch...
--->14:57:42.224 MQT: verbunden                                                                 ~~~11.2 secs
14:57:42.228 MQT: tele/tasmota_xxxxxx/LWT = Online (beibehalten)
14:57:42.261 MQT: stat/tasmota_xxxxxx/MQTTRETRY = {"MqttRetry":10}   > Set to 10 again via MQTT
14:57:42.312 MQT: stat/tasmota_xxxxxx/POWER = {"POWER":"OFF"}
```
